File: gender.py
"""Infer product gender/audience from catalog naming conventions."""

from __future__ import annotations

import re

# (?<![a-z]) prevents "men" matching inside "women" / "amen".
_GENDER_RULES: tuple[tuple[re.Pattern[str], str], ...] = (
    (re.compile(r"\bunisex\b", re.IGNORECASE), "unisex"),
    (re.compile(r"\bwomen(?:'s|s)?\b", re.IGNORECASE), "women"),
    (re.compile(r"(?<![a-z])men(?:'s|s)?\b", re.IGNORECASE), "men"),
    (re.compile(r"\bkids?\b", re.IGNORECASE), "kids"),
    (re.compile(r"\bboys?\b", re.IGNORECASE), "kids"),
    (re.compile(r"\bgirls?\b", re.IGNORECASE), "kids"),
)

_VALID_GENDERS = frozenset({"men", "women", "kids", "unisex"})


def normalize_gender(value: str | None) -> str:
    gender = (value or "").strip().lower()
    return gender if gender in _VALID_GENDERS else ""


def infer_gender(name: str, brand: str = "") -> str:
    """Return men | women | kids | unisex | empty string if unknown."""
    if not name:
        return ""

    haystack = name
    if brand:
        prefix = brand.strip()
        if name.lower().startswith(prefix.lower()):
            haystack = name[len(prefix) :].strip()

    for pattern, gender in _GENDER_RULES:
        if pattern.search(haystack) or pattern.search(name):
            return gender
    return ""


def resolve_product_gender(
    *,
    name: str,
    brand: str = "",
    stored_gender: str | None = None,
) -> str:
    """Prefer explicit name tokens over a stale stored gender value."""
    inferred = infer_gender(name, brand)
    stored = normalize_gender(stored_gender)
    if inferred and stored and inferred != stored:
        return inferred
    return stored or inferred

File: web/lib/gender.ts
export type GenderValue = "men" | "women" | "kids" | "unisex";

const VALID_GENDERS = new Set<GenderValue>(["men", "women", "kids", "unisex"]);

export function normalizeGender(value?: string): GenderValue | "" {
  const gender = (value || "").trim().toLowerCase();
  return VALID_GENDERS.has(gender as GenderValue) ? (gender as GenderValue) : "";
}

export function inferGender(name: string, brand = ""): string {
  let haystack = name;
  if (brand && name.toLowerCase().startsWith(brand.toLowerCase())) {
    haystack = name.slice(brand.length).trim();
  }

  const rules: [RegExp, string][] = [
    [/\bunisex\b/i, "unisex"],
    [/\bwomen(?:'s|s)?\b/i, "women"],
    [/(?<![a-z])men(?:'s|s)?\b/i, "men"],
    [/\bkids?\b/i, "kids"],
    [/\bboys?\b/i, "kids"],
    [/\bgirls?\b/i, "kids"],
  ];

  for (const [pattern, gender] of rules) {
    if (pattern.test(haystack) || pattern.test(name)) return gender;
  }
  return "";
}

export function productGender(
  product: { name: string; brand: string; gender?: string },
): string {
  const inferred = inferGender(product.name, product.brand);
  const stored = normalizeGender(product.gender);
  if (inferred && stored && inferred !== stored) {
    return inferred;
  }
  return stored || inferred;
}
